submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,197 @@
+## Globalnet 2.0: Enhancements to Globalnet Implementation
+
+## Summary
+
+Currently, Submariner Globalnet implementation in [release 0.8](https://github.com/submariner-io/releases/releases/tag/v0.8.0)
+annotates every Pod, Service and Node in the cluster with a globalIp to support Pod to remote
+Service connectivity in the following manner.
+
+Once a Pod is scheduled and enters Running state, it is annotated with a globalIp. To support
+the connectivity, Globalnet programs the necessary egress rules on the active Gateway Node of
+the cluster where the source IP of the Pod is modified to the corresponding globalIp. This
+translation enables us to uniquely identify the traffic from the Pod, when it reaches the
+destination cluster.
+
+Similarly, a ClusterIP Service is annotated with a globalIp so that remote Pods can communicate
+with it using its globalIp. Globalnet programs the necessary ingress rules on the active Gateway
+node of the cluster, where the service resides. The ingress rules enable any incoming traffic
+destined to the globalIp of the Service to be redirected to the corresponding ClusterIP Service.
+
+In order to support connectivity from the Host Network to the remote services, Globalnet also
+annotates every node in the cluster with a unique globalIp.
+
+While this works fine for small clusters, where the number of Pods and Services are limited, for
+large clusters it poses scalability and performance issues as we would eventually run out of
+globalIps because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod
+connectivity as well as Headless Services are not supported.
+
+## Proposal
+
+This enhancement proposes the following changes to make Globalnet implementation scale better.
+
+1. By default, every cluster will be assigned a single globalIp which will be used as egress IP
+   for any cross-cluster communication.
+2. Provision to allocate a globalIp at the namespace/project level which takes precedence over
+   the globalIp allocated to the cluster.
+3. Provision to allocate a globalIp to a Pod (or a set of Pods) which has higher precedence
```

Comment:
I feel like at the pod level (and probably this can apply to namespace level as well) explicit is better than implicit, so it seems to me we can have the "global ip policy" annotate the pods with the global IP.

This will give the user visibility which global IP is used for which pod which can be great for debugging.

This is only for pods expliditly selected, if pod is not selected it will fall back to namespace/cluster policy.

I think it'll also make the whole thing very easy for us to compute.

EG when defining rules on the gateway we can put pod rules first, then catch all rules for namespaces, and finally the rule for the cluster.

Reply:
yup, it could be done in terms of iptables with chains, you have a chain for pods, a chain for namespaces, a general chain as a fallback, @mkimura suggested this.

Reply:
Yes, this is possible and we already do something similar in the current Globalnet implementation. We first send packets to SUBMARINER-GN-MARK chain to mark packets destined to remoteClusters and in the subsequent chains (which are like fallback chains) we use the mark to make decisions.